File: Algorithms/1339_Maximum_Product_of_Splitted_Binary_Tree/Java/MaximumProductofSplittedBinaryTreeSolution1.java

// Space: O(n)
// Time: O(n)

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private Map<TreeNode, Long> cache = new HashMap<>();

    public int maxProduct(TreeNode root) {
        dfs(root);
        int mod = 1000000007;
        long totalSum = cache.get(root), res = 0;
        Queue<TreeNode> queue = new ArrayDeque<>();
        queue.add(root);
        while (!queue.isEmpty()) {
            TreeNode cur = queue.poll();
            if (cur.left != null) {
                res = Math.max(res, (totalSum - cache.get(cur.left)) * cache.get(cur.left));
                queue.add(cur.left);
            }
            if (cur.right != null) {
                res = Math.max(res, (totalSum - cache.get(cur.right)) * cache.get(cur.right));
                queue.add(cur.right);
            }
        }
        return (int) (res % mod);
    }

    private long dfs(TreeNode root) {
        if (root == null) return 0;
        if (cache.containsKey(root)) return cache.get(root);
        cache.put(root, dfs(root.left) + dfs(root.right) + root.val);
        return cache.get(root);
    }
}
         

File: Algorithms/1339_Maximum_Product_of_Splitted_Binary_Tree/Python/Maximum_Product_of_Splitted_Binary_Tree_Solution_1.py

# Space: O(n)
# Time: O(n)

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def maxProduct(self, root):
        cache = {}
        self.res = 0
        mod = 1000000007

        def dfs(root):
            if root is None: return 0
            if root in cache: return cache[root]
            left = dfs(root.left)
            right = dfs(root.right)
            cache[root] = root.val + left + right
            return cache[root]

        dfs(root)

        queue = [root]
        while queue:
            cur = queue.pop(0)
            if cur.left:
                self.res = max(self.res, (cache[root] - cache[cur.left]) * cache[cur.left])
                queue.append(cur.left)
            if cur.right:
                self.res = max(self.res, (cache[root] - cache[cur.right]) * cache[cur.right])
                queue.append(cur.right)
        return self.res % mod




